feat(rust/sedona-geo): Add ST_AsGeoJSON

prantogg opened the pull request:
Implements the `ST_AsGeoJSON` function for converting geometry objects to GeoJSON format.

petern48 commented on rust/sedona-functions/src/st_asgeojson.rs:
```diff
@@ -0,0 +1,456 @@
+}
+
+#[cfg(test)]
+mod tests {
```
In addition to rust tests, we need integration tests in [test_functions.py](https://github.com/apache/sedona-db/blob/e0e1d109480727faaf7be25923b57b4686144438/python/sedonadb/tests/functions/test_functions.py#L4). Instructions for testing it locally and iterating are [here](https://github.com/apache/sedona-db/blob/e0e1d109480727faaf7be25923b57b4686144438/docs/contributors-guide.md#python).

petern48 commented on rust/sedona-functions/src/st_asgeojson.rs:
```diff
@@ -0,0 +1,456 @@
+    args: &[ColumnarValue],
+    geojson_type: GeoJsonType,
+) -> Result<ColumnarValue> {
+    let executor = WkbExecutor::new(arg_types, args);
```
+1 to moving to this from `rust/sedona-functions/` to `rust/sedona-geo`. Note you need to do this *in order to* access these helpful methods/structs.

`GeoTypesExecutor` uses a [item_to_geometry()](https://github.com/apache/sedona-db/blob/e0e1d109480727faaf7be25923b57b4686144438/rust/sedona-geo/src/to_geo.rs#L58) function that implements exactly what you did in `wkb_to_geometry()` below (hence we can remove it). If you find `GeoTypesExecutor` isn't flexible enough for this function ([like I did](https://github.com/apache/sedona-db/pull/233#discussion_r2443552552) for a different function), you could consider using `item_to_geometry()` directly instead.

petern48 commented on rust/sedona-functions/src/st_asgeojson.rs:
```diff
@@ -0,0 +1,67 @@
+    )
+    .with_sql_example("SELECT ST_AsGeoJSON(ST_Point(1.0, 2.0))")
+    .with_sql_example("SELECT ST_AsGeoJSON(ST_Point(1.0, 2.0), 'Feature')")
+    .with_sql_example("SELECT ST_AsGeoJSON(ST_Point(1.0, 2.0), 'FeatureCollection')")
```
```suggestion
```

petern48 replied:
```suggestion
```

prantogg commented:
Thanks for the feedback! I've moved the code to `rust/sedona-geo` and am now using `GeoTypesExecutor` as suggested. I've also removed the output type parameter for now - we can address that in a later PR.

However, I'm running into some [CI python test failures](https://github.com/apache/sedona-db/actions/runs/20584413572/job/59118209352?pr=469#step:14:3679) when comparing GeoJSON output between SedonaDB and PostGIS.
The differences arise from both the geojson crate's serialization and how it converts from `geo_types::Geometry`. Specifically:
- Coordinate formatting (integers vs floats): The `geojson` crate always serializes coordinates as floats, even for whole numbers. PostGIS optimizes by omitting .0 for whole numbers.
- Empty polygon representation: The geojson crate's conversion from `geo_types::Polygon` to `geojson::Geometry` represents `POLYGON EMPTY` as `{"type":"Polygon","coordinates":[[]]}` (one empty ring), while PostGIS returns `{"type":"Polygon","coordinates":[]}` (no rings).

What's your preference for handling these format differences? Should we:
- Normalize these in the tests (converting PostGIS output to match SedonaDB's format)?
- Look into customizing the GeoJSON serialization to match PostGIS exactly?

petern48 commented on rust/sedona-geo/src/st_asgeojson.rs:
```diff
@@ -0,0 +1,149 @@
+
+        // Estimate the minimum probable memory requirement of the output.
+        // GeoJSON is typically longer than WKT due to JSON formatting.
+        let min_probable_geojson_size = executor.num_iterations() * 50;
```
```suggestion
        // Minimal GeoJSON: {"type":"Point","coordinates":[]}
        let min_probable_geojson_size = executor.num_iterations() * 33;
```
how about this?

petern48 reviewed (commented):
> The geojson crate's conversion from geo_types::Polygon to geojson::Geometry represents POLYGON EMPTY as {"type":"Polygon","coordinates":[[]]} (one empty ring), while PostGIS returns {"type":"Polygon","coordinates":[]} (no rings).

I do think this is technically a bug upstream (see the toggle for why), and there does seem to be a maintainer responding to new isssues/PRs in the [geojson rust repo](https://github.com/georust/geojson). I'd say if you're eager to fix it, then go for it, but I think it's fine to kick this down the road since this is such a specific edge case. At least it's valid traditional JSON. We can just file an issue and comment the test out.

<details>
<summary>Proof from the [GeoJSON spec](https://datatracker.ietf.org/doc/html/rfc7946) (I asked an LLM and verified it 🤠)</summary>

`A linear ring is a closed LineString with four or more positions.`

`For type "Polygon", the "coordinates" member MUST be an array of linear ring coordinate arrays.`

</details>

> Coordinate formatting (integers vs floats): The geojson crate always serializes coordinates as floats, even for whole numbers. PostGIS optimizes by omitting .0 for whole numbers.

I think fixing this would be convenient to do when we decide to add support for the `integer maxdecimaldigits` argument for `ST_AsGeoJSON`, which would likely involve modifying similar code upstream. Speaking of, it seems like there's an old PR that might be fixing this for us: https://github.com/georust/geojson/pull/234, though I didn't look too deeply into it.

For both cases, I'm in favor of tweaking the tests a bit to pass and keeping existing behavior for convenience. We'll see what Dewey thinks.

petern48 commented on rust/sedona-functions/src/st_asgeojson.rs:
```diff
@@ -0,0 +1,456 @@
+                                "Failed to convert to GeoJSON: {:?}",
+                                e
+                            ))
+                        })?;
```
filed a ticket so we don't forget: https://github.com/apache/sedona-db/issues/472
